controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,684 @@
+				PublicKeys: &[]containerservice.SSHPublicKey{
+					{
+						KeyData: to.StringPtr(spec.LinuxSSHPublicKeyContents),
+					},
+				},
+			},
+		}
+	}
+
+	managedCluster := containerservice.ManagedCluster{
+		Name:		to.StringPtr(spec.ClusterName),
+		Location: 	to.StringPtr(spec.ResourceLocation),
+		Tags:     	tags,
+		ManagedClusterProperties: &containerservice.ManagedClusterProperties{
+			KubernetesVersion: to.StringPtr(kubernetesVersion),
+			DNSPrefix:         to.StringPtr(dnsPrefix),
+			AgentPoolProfiles: &agentPoolProfiles,
+			LinuxProfile:      linuxProfile,
+			NetworkProfile:    networkProfile,
+			ServicePrincipalProfile: &containerservice.ManagedClusterServicePrincipalProfile{
+				ClientID: to.StringPtr(spec.ClientID),
+				Secret:   to.StringPtr(spec.ClientSecret),
+			},
+		},
+	}
+
+	future, err := clusterClient.CreateOrUpdate(
+		ctx,
+		spec.ResourceGroup,
+		spec.ClusterName,
+		managedCluster,
+	)
+	if err != nil {
+		return c, fmt.Errorf("[AKS] cannot create AKS cluster: %v", err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, clusterClient.Client); err != nil {
+		return c, fmt.Errorf("[AKS] cannot get the AKS cluster create or update future response: %v", err)
+	}
+
+	return future.Result(*clusterClient)
+}
+
+// Check if AKS managed Kubernetes cluster exist
+func checkCluster(ctx context.Context, clusterClient *containerservice.ManagedClustersClient, spec aksv1.AKSClusterConfigSpec) (bool, error) {
+	resp, err := clusterClient.Get(ctx, spec.ResourceGroup, spec.ClusterName)
+
+	if err != nil {
+		return false, fmt.Errorf("[AKS] error getting resource group '%s': %v", spec.ResourceGroup, err)
+	}
+
+	return resp.StatusCode == 200, nil
+}
+
+// Delete AKS managed Kubernetes cluster
+func removeCluster(ctx context.Context, clusterClient *containerservice.ManagedClustersClient, spec aksv1.AKSClusterConfigSpec) (err error) {
+	future, err := clusterClient.Delete(ctx, spec.ResourceGroup, spec.ClusterName)
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletionRef(ctx, clusterClient.Client)
+	if err != nil {
+		logrus.Errorf("[AKS] cannot get the AKS cluster create or update future response: %v", err)
+		return err
+	}
+
+	logrus.Infof("Cluster %v removed successfully", spec.ClusterName)
+	logrus.Infof("Cluster removal status %v", future.Status())
+
+	return nil
+}
+
+func hasCustomVirtualNetwork(spec aksv1.AKSClusterConfigSpec) bool {
+	return spec.VirtualNetwork != "" && spec.Subnet != ""
+}
+
+func hasAgentPoolProfile(spec aksv1.AKSClusterConfigSpec) bool {
+	return spec.NodeGroups != nil
+}
+
+func hasLinuxProfile(spec aksv1.AKSClusterConfigSpec) bool {
+	return spec.LinuxAdminUsername != "" && (spec.LinuxSSHPublicKeyContents != "")
+}
+
+func isKubernetesVersion(version string) bool {
+    switch version {
+    case
+    	"1.19.6",
+		"1.19.3",
+		"1.18.14",
+		"1.18.10",
+		"1.17.16",
+		"1.17.13",
+		"1.16.15",
+		"1.16.13":
+        return true
+    }
+    return false
+}
```

Comment:
If you use an improper version and send it to AKS will an error be returned? If so we should depend on that and not put ourselves in a situation where version options are hard coded. We can however check validate general input rules.

Reply:
This was changed and currently it is validated against Azure API